### PR TITLE
[SYCLomatic] Fix unsuccessful compilation of `dpct::load_kernel_library_mem` due to failed argument cast

### PR DESCRIPTION
**Pull request by the-slow-one**
Here is the CUDA source and migrated SYCL code.

```
void test(CUmodule module, unsigned char* data) {
  cuModuleLoadData(&module, data);
}
```

```
#include <sycl/sycl.hpp>                                                                                                                                                                                                                                                                           #include <dpct/dpct.hpp>
void test(dpct::kernel_library module, unsigned char* data) {
  /*
  DPCT1104:0: 'data' should point to a dynamic library loaded in memory. The
  dynamic library should supply wrapped kernel functions.
  */
  module = dpct::load_kernel_library_mem(data);
}
```

`icpx` fails to compile the migrated code with following error.
```
dpct_output/src.dp.cpp:11:42: error: cannot initialize a parameter of type 'const char *' with an lvalue of type 'u8 *' (aka 'unsigned char *')
11 |   module = dpct::load_kernel_library_mem(data);
   |                                          ^~~~
/nfs/shm/proj/icl/cmplrarch/deploy_syclct/llorgsyclctefi2linux/20241125_160000/build/linux_qa_release/include/dpct/kernel.hpp:376:72: note: passing argument to parameter 'image' here
376 | static inline kernel_library load_kernel_library_mem(char const *const image) {
    |                                                                        ^
1 error generated.
```

This PR fixes this issue by explicitly casting the pointer the expected type.

Signed-off-by: [Deepak Raj H R](https://github.com/oneapi-src/SYCLomatic/pull/Deepak.raj.h.r@intel.com)

### CODE REVIEW

File: clang/lib/DPCT/RulesLang/APINamesDriver.inc [hunk 1]
```diff
@@ -26,14 +26,14 @@ FEATURE_REQUEST_FACTORY(
     ASSIGNABLE_FACTORY(ASSIGN_FACTORY_ENTRY("cuModuleLoadData", DEREF(0),
                                             CALL(MapNames::getDpctNamespace() +
                                                      "load_kernel_library_mem",
-                                                 ARG(1)))))
+                                                 CAST(makeLiteral("const char *"), ARG(1))))))
```
Comment (zhiweij1) [inline clang/lib/DPCT/RulesLang/APINamesDriver.inc]: How about
```suggestion
                                                 CAST_IF_NOT_SAME(makeLiteral("const char *"), ARG(1))))))
```
?

File: clang/lib/DPCT/RulesLang/APINamesDriver.inc [hunk 2]
```diff
@@ -26,14 +26,14 @@ FEATURE_REQUEST_FACTORY(
...
 
 FEATURE_REQUEST_FACTORY(
     HelperFeatureEnum::device_ext,
     ASSIGNABLE_FACTORY(ASSIGN_FACTORY_ENTRY("cuModuleLoadDataEx", DEREF(0),
                                             CALL(MapNames::getDpctNamespace() +
                                                      "load_kernel_library_mem",
-                                                 ARG(1)))))
+                                                 CAST(makeLiteral("const char *"), ARG(1))))))
```
Comment (zhiweij1) [inline clang/lib/DPCT/RulesLang/APINamesDriver.inc]: Same as above

Comment (the-slow-one) [inline clang/lib/DPCT/RulesLang/APINamesDriver.inc]:
```suggestion
                                                 CAST_IF_NOT_SAME(makeLiteral("const char *"), ARG(1))))))
```

File: clang/test/dpct/module_main.cu [hunk 3]
```diff
@@ -20,15 +20,28 @@ int main(){
     //CHECK: /*
     //CHECK-NEXT: DPCT1104:{{[0-9]+}}: 'Data.c_str()' should point to a dynamic library loaded in memory. The dynamic library should supply wrapped kernel functions.
     //CHECK-NEXT: */
-    //CHECK-NEXT: M = dpct::load_kernel_library_mem(Data.c_str());
+    //CHECK-NEXT: M = dpct::load_kernel_library_mem((const char *)Data.c_str());
```
Comment (the-slow-one) [inline clang/test/dpct/module_main.cu]:
```suggestion
    //CHECK-NEXT: M = dpct::load_kernel_library_mem(Data.c_str());
```

File: clang/test/dpct/module_main.cu [hunk 4]
```diff
@@ -20,15 +20,28 @@ int main(){
...
+    //CHECK-NEXT: DPCT1104:{{[0-9]+}}: 'DataPtr)' should point to a dynamic library loaded in memory. The dynamic library should supply wrapped kernel functions.
+    //CHECK-NEXT: */
+    //CHECK-NEXT: M = dpct::load_kernel_library_mem((const char *)DataPtr);
+    cuModuleLoadData(&M, DataPtr);
+
     //CHECK: /*
     //CHECK-NEXT: DPCT1104:{{[0-9]+}}: 'Data.c_str()' should point to a dynamic library loaded in memory. The dynamic library should supply wrapped kernel functions.
     //CHECK-NEXT: */
-    //CHECK-NEXT: M = dpct::load_kernel_library_mem(Data.c_str());
+    //CHECK-NEXT: M = dpct::load_kernel_library_mem((const char *)Data.c_str());
```
Comment (the-slow-one) [inline clang/test/dpct/module_main.cu]:
```suggestion
    //CHECK-NEXT: M = dpct::load_kernel_library_mem(Data.c_str());
```

**Review by zhiweij1 — APPROVED** [review]

**Review by TejaX-Alaghari — APPROVED** [review]

**Review by ziranzha — APPROVED** [review]

**Pull request merged**